tsfc/loopy.py

```diff
@@ -40,26 +43,67 @@ def pym_multiindex(self, multiindex):
                 indices.append(index)
         return tuple(indices)
 
-    def pymbolic_variable(self, node):
+    def gem_to_pym(self, node):
         try:
             pym = self.gem_to_pymbolic[node]
         except KeyError:
             name = self.name_gen(node.name)
             pym = p.Variable(name)
             self.gem_to_pymbolic[node] = pym
+        return pym
+
+    def pymbolic_variable(self, node):
+        pym = self.gem_to_pym(node)
         if node in self.indices:
             indices = self.pym_multiindex(self.indices[node])
             if indices:
                 return p.Subscript(pym, indices)
-            else:
-                return pym
-        else:
-            return pym
+        return pym
+
+    def pymbolic_variable_and_destruct(self, node):
+        pym = self.gem_to_pym(node)
+        if node in self.indices:
+            indices = self.pym_multiindex(self.indices[node])
+            if indices:
+                return (pym, indices)
+        return (pym, ())
 
     def active_inames(self):
         # Return all active indices
         return frozenset([i.name for i in self.active_indices.values()])
 
+    def pymbolic_indices(self, expr):
+        try:
+            indices = expr.multiindex
+        except AttributeError:
+            indices = tuple(shp for shp in expr.shape)
+        pym_idx = tuple()
+        for index in indices:
+            if type(index) == gem.Index:
+                if not index.name:
+                    name = self.name_gen(self.index_names[index])
+                else:
+                    name = index.name
+                assert index.extent
+                extent = index.extent
+            else:
+                name = self.name_gen(self.index_names[index])
+                extent = index
+            self.index_extent[name] = extent
+            pym_idx += (p.Variable(name), )
+        return pym_idx
```

Comment:
So this is to generate index names and extents for some expression. It looks like it is only ever called on shaped objects, but one of them `ComponentTensor` has bound its indices already, and the others don't have indices in the gem representation. So does it not make more sense to use `pym_multiindex` in the component tensor code path and just use the shape-based code path for the other cases?

Reply:
Pym_multiindex does not generate the pymbolic Variable for the index at the moment. Do you just want me to move the code for the ComponentTensor case to pym_multiindex?

Reply:
So I think if you do that, then it's clearer in the Solve and Inverse case that you're making new indices rather than turning gem indices into pymbolic ones?